netsim/cli/exec.py

```diff
@@ -0,0 +1,104 @@
+#
+# netlab exec command
+#
+# run a command on one or more lab devices
+#
+import typing
+import os
+import sys
+import argparse
+import subprocess
+from enum import IntEnum
+
+from box import Box
+
+from . import external_commands, set_dry_run
+from . import load_snapshot, _nodeset, parser_add_verbose
+from ..outputs import common as outputs_common
+from ..utils import strings, log
+
+from .connect import quote_list, docker_connect, ssh_connect, LogLevel, get_log_level
+
+
+
+#
+# CLI parser for 'netlab ' command
+#
+def connect_parse(args: typing.List[str]) -> typing.Tuple[argparse.Namespace, typing.List[str]]:
+  parser = argparse.ArgumentParser(
+      prog="netlab exec",
+      description='Run a command on one or more network devices',
+      epilog='The rest of the arguments are passed to SSH or docker exec command')
+  parser_add_verbose(parser)
+  parser.add_argument(
+      '--dry-run',
+      dest='dry_run',
+      action='store_true',
+      help='Print the hosts and the commands that would be executed on them, but do not execute them')
+  parser.add_argument(
+      '--snapshot',
+      dest='snapshot',
+      action='store',
+      nargs='?',
+      default='netlab.snapshot.yml',
+      const='netlab.snapshot.yml',
+      help='Transformed topology snapshot file')
+  parser.add_argument(
+      dest='node', action='store',
+      help='Node(s) to run command on')
+  return parser.parse_known_args(args)
+
+       
+def exec_on_node(
+      args: argparse.Namespace,
+      rest: list,
+      topology: Box,
+      node: str,
+      log_level: LogLevel = LogLevel.INFO) -> typing.Union[bool,int,str]:
+  
+  host_data = outputs_common.adjust_inventory_host(
+                node=topology.nodes[node],
+                defaults=topology.defaults,
+                group_vars=True)
+  host_data.host = node
+  connection = host_data.netlab_console_connection or host_data.ansible_connection
+
+  if connection == 'docker':
+    return docker_connect(host_data,args,rest,log_level)
+  elif connection in ['paramiko','ssh','network_cli','netconf','httpapi'] or not connection:
+    if connection in ['netconf','httpapi']:
+      print(f"Using SSH to connect to a device configured with {connection} connection")
+    return ssh_connect(host_data,args,rest,log_level)
+  else:
+    log.fatal(f'Unknown connection method {connection} for host {node}',module='connect')
+
+
+def run(cli_args: typing.List[str]) -> None:
+  (args, rest) = connect_parse(cli_args)
```

Comment:
You should test that there are some extra arguments, otherwise the command will interactively connect to the every box in the nodeset.